More fixes

DisasterMoo opened the pull request:
- Blast Furnace multiblock structure allows Tier III and above metal sheets (Closes #1259).
- Fixed small vessel dupe in crucible (Closes #1258).
- Added salting recipe wrapper to JEI and fixed log output (Closes #1018).
- Added tree growth to waila provider (Closes #1262).
- Crops now require sky and moisturized farmland, otherwise it will reset growth to the start of current stage (Closes #342, Closes #1249).
- Fixed animals dying regardless of age, also set default config back to disabled (Closes #1268).
- Changed cooked eggs to spoil faster and made raw eggs be the choice for preserving it (Closes #1265)
- Milking now ignore creative and fills the bucket (since it drains from the cow/goat and they have a cooldown, should stop people thinking it is a bug)
- Added missing mating AI to get mules
- Fixed issue in hwyla integration (Closes #1223).
- Fixed crash in chested horses (Closes #1272).
- Made some blocks drop itself if not "supported" (Closes #1154, Closes #1165).
- Fixed quern hitboxes not rendering when looked upon above (Closes #1017).
- Skeleton horses will not spawn on surface anymore (Closes #1248).
- Fixed animals bounding boxes (Closes #1229).
- Fixed torches being able to be placed on large vessels (Closes #1275).
- Added hunting lists to carnivores (Closes #1022, Closes #1073).
- Familiarity overlay now renders only on familiarizable animals (Closes #1099).
- Added food lists to livestock animals (Closes #1240).
- Added convenient support placement on sneaking right click (Closes #1237).
- Added config to change mining time of rock and log blocks (Closes #1060).
- Added CT integration to bloomery and blast furnace (finally, Closes #888).
- Knapping GUI now uses the processing arrow instead of a question mark for JEI click areas (Closes #907)
- Fixed goats being able to be milked without cooldown (Closes #1277).
- Refactored vinegar preservation to use a barrel recipe instead of fluid property. Also, vinegar preservation now requires 125mB of vinegar for each food piece (Closes #1266).
- Rains now make fire pits and charcoal forges consumes more fuel (Closes #1276).
- TFC foods will not decay in any creative tab and in JEI (Closes #1109).
- Fertilized eggs' tooltip now works! (Closes #1234).
- Added gem[Quality][gemName] ore tag to gems (Closes #1148).
- Food poisoning now has it's own poison effect (Closes #996).
- Barrels now uses a fluid whitelist (Closes #942)

DisasterMoo committed:
Changed blast furnace sheets to allow any Tier III and above metals (Closes #1259).

DisasterMoo committed:
Fixed small vessel dupe in crucible (Closes #1258).

DisasterMoo committed:
Added salting recipe wrapper to JEI (Closes #1018). Also fixed both salting and unmold recipes error-ing in log.

DisasterMoo committed:
Added tree growth to waila provider (Closes #1262).

DisasterMoo committed:
Crops now requires farmland to be moisturized (a little tiny bit at least) and to see sky otherwise it will reset growth to the start of current stage *evil laughter* (Closes #1249, Closes #342).

alcatrazEscapee commented on src/main/java/net/dries007/tfc/compat/jei/wrappers/SaltingRecipeWrapper.java:
```diff
@@ -0,0 +1,68 @@
+                {
+                    // Reset capability to avoid rotten foods
+                    food.setCreationDate(CalendarTFC.PLAYER_TIME.getTicks());
+                    food.getTraits().clear();
```
Should not have to do this. This is either
 - loaded before a world exists, which means no item stacks created will decay (and also `CalendarTFC.PLAYER_TIME` is meaningless, or
 - should be replaced with a `food.setNonDecaying()`

DisasterMoo replied:
Yeah, my bad. I didn't notice the first time it was not decaying by default. Now what i found strange is that it is coming with brined from `Ingredient#getMatchingStacks`. But nonetheless, the recipe is salting non salted foods as expected, so i left `food.getTraits().clear()` there.

DisasterMoo committed:
Changed default old animals death chance to 0 (disabled) by default. Also, fixed PR changes.

DisasterMoo committed:
A few changes:
- Fixed a animals being selected to die at any age (Closes #1268).
- Lowered raw eggs' decay rate (should last for about 16 days, Closes #1265).
- Made cooked eggs rot faster to match (mechanic here should be keep raw eggs for duration).
- Milking now ignore creative and fill the bucket anyway.

DisasterMoo committed:
A few more changes:
- Added missing breeding AI to get mules in horses and donkeys.
- Fixed an issue in hwyla integration (Closes #1223).
- Fixed crash in chested horse-like entities (Closes #1272).

DisasterMoo committed:
Made nest boxes, lamps and large vessels drop if unsupported (Closes #1154, Closes #1165).

DisasterMoo committed:
Fixed #1017.

alcatrazEscapee commented on src/main/java/net/dries007/tfc/objects/blocks/devices/BlockBlastFurnace.java:
```diff
@@ -47,29 +47,35 @@
+            if (state.getBlock() instanceof BlockMetalSheet)
+            {
+                BlockMetalSheet block = (BlockMetalSheet) state.getBlock();
+                return block.getMetal().getTier().isAtLeast(Metal.Tier.TIER_III);
```
One thing to consider: As is, this allows metals that are technically "unfinished products", i.e. high carbon steels, weak steels, pig iron, etc. This *may* be fixed, likely just by requiring this as to also be `isToolMetal()` (which would restrict the list in TFC to iron, steel, or colored steel). However I'm unsure of if this change should actually be made

DisasterMoo committed:
Cleanup

DisasterMoo committed:
Fixed #1248.

DisasterMoo committed:
Animals now update their bounding boxes once every 5 secs, Client and Server side (Closes #1229).

DisasterMoo committed:
Fixed #1275.

DisasterMoo committed:
Added creature hunting list to carnivore animals (Closes #1022, Closes #1073).

DisasterMoo committed:
Familiarity overlay only renders if animal is familiarizable (Closes #1099).

DisasterMoo committed:
Added configuration for what livestock animals accept as food (Closes #1240).

DisasterMoo committed:
Added convenient placement of support blocks on top if you right click while sneaking (Closes #1237).

DisasterMoo committed:
Added config to change mining time of rock and log blocks (Closes #1060).

DisasterMoo committed:
Added CT hooks for blast furnace and bloomery (Closes #888). Also, minor nitpick fix on blast furnace recipes (registry name has changed).

DisasterMoo committed:
Knapping GUI now uses the processing arrow instead of a question mark for JEI recipe search click area (only one that needed this change, Closes #907).

DisasterMoo committed:
Fixed goats being able to be milked without cooldown (Closes #1277).

DisasterMoo committed:
Rains now turn off fire pits and charcoal forges (Closes #1276).

alcatrazEscapee commented on src/main/java/net/dries007/tfc/objects/entity/animal/EntityGoatTFC.java:
```diff
@@ -70,6 +75,61 @@ public void birthChildren()
+            return false;
+        }
+        // Check if item is accepted
+        for (String input : ConfigTFC.Animals.GOAT.food)
```
Firstly, this is *horrendously* gross. It's also duplicated verbatim all over various animals, which just makes it worse. It's using what looks to be a completely custom format (which I don't see a formal spec for anywhere), and you're also doing the parsing of the actual ingredient on demand, every time.

It's a hodgepodge of formatting and conventions:
 - craft tweaker esque "ore" syntax that has no business here.
 - checking for metadata `-1` values, instead of `OreDictionary.WILDCARD_VALUE` or other wildcards.
 - space seperated metadata and item name

An alternate proposal:
 - Change the config to a single string.
 - The format of this string must be a valid `Ingredient` (vanilla) json. (although obviously newlines removed because it's a config and all of these config values are hacked onto what really should be (and will be) json eventually).
 - Then, to parse this, use `gson` and call `CraftingHelper#getIngredient(JsonElement, JsonContext)` to return a vanilla `Ingredient` instance. Lazily cache this as part of the animal data, and invalidate it when configs change (there's an event for that).

DisasterMoo replied:
Copy from discord:

> I concede the first implementation i did for animal feeding list is gross, but putting a json in a config? Don't seem right to me, if we want to use json (and i prefer to do that in 1.15 for certain, although i would use the datapacks if possible), I prefer to use an actual json file, like i did for damage resistances
so, the counter-proposal is that, intead of a config, a json file containing all current animals food with a loader=animalFoodList

Thoughts?

DisasterMoo replied:
Done!

alcatrazEscapee commented on src/main/java/net/dries007/tfc/objects/te/TECharcoalForge.java:
```diff
@@ -87,6 +87,15 @@ public void update()
             IBlockState state = world.getBlockState(pos);
             if (state.getValue(LIT))
             {
+                // Have to check the above block, since minecraft think this block is "roof"
+                if (world.isRainingAt(pos.up()))
+                {
+                    // Instantly consume last fuel and turn off
+                    burnTicks = 0;
+                    burnTemperature = 0;
+                    world.setBlockState(pos, state.withProperty(LIT, false));
+                    return;
+                }
```
Is there a particular reason this both 
 - checks every tick for rain instead of on random ticks or % chances, and
 - has a 100% chance to go out on said exact tick?

Given the charcoal forge has a requirement that it must see sky light (although it can be indirect), I would assume a much lower chance of going out in the rain. It's also a charcoal bleedin forge that can reach temperatures > 1500 C, but rain puts it out *instantly*?

alcatrazEscapee commented on src/main/java/net/dries007/tfc/objects/te/TEFirePit.java:
```diff
@@ -125,6 +125,15 @@ public void update()
             IBlockState state = world.getBlockState(pos);
             if (state.getValue(LIT))
             {
+                // Have to check the above block, since minecraft think this block is "roof"
+                if (world.isRainingAt(pos.up()))
+                {
+                    // Instantly consume last fuel and turn off
+                    burnTicks = 0;
+                    burnTemperature = 0;
+                    world.setBlockState(pos, state.withProperty(LIT, false));
+                    return;
+                }
```
Same as above, is there a reason this isn't done on random tick? That would make more sense thematically.

DisasterMoo committed:
Refactored vinegar preservation to use a barrel recipe instead of fluid property. Also, vinegar preservation now requires 125mB of vinegar for each food piece (Closes #1266).

DisasterMoo committed:
A few changes for review:
- Rain now removes a handful (configurable) number of burn ticks from charcoal forge and fire pits on randomTicks
- Blast furnace multiblock structure now requires tool metals of tier III and above.

DisasterMoo committed:
Fix food decaying in JEI and creative tabs once and for all (Applies only to TFC foods, Closes #1109).

DisasterMoo committed:
Json empire strikes back! (animal food list now in json instead of config).

DisasterMoo committed:
Fixed #1234.

DisasterMoo committed:
Added more ore dictionary tags to gems (Closes #1148).

DisasterMoo committed:
Food poisoning now has it's own potion effect (Closes #996). Also, made a use for thirst debuff potion effect by applying when player drinks salty water (chance, configurable).

DisasterMoo committed:
Barrels now only accepts fluids that are registered as ingredient for fluids or in a configurable whitelist (Closes #942).

DisasterMoo committed:
Config here, config there, another one here... Damn you configs! (Closes #1145, Closes #1111, Closes #1108, Closes #1106, Closes #1105, Closes #934).

DisasterMoo committed:
Explosions now cause cave-ins (with config to disable, Closes #892).

DisasterMoo committed:
Added stone pressure plates and recipes (Closes #1096). Also, fixed armor stand recipe (Closes #1283).
